feat(dot/parachain): implement validate candidate from exhaustive

edwardmack opened the pull request:
## Changes
This PR introduces code for candidate validation subsystem to validate candidates when all data validation data is provided.  As described in [Implementation Plan ](https://hackmd.io/p-6njHUST8uSlNdKRUUHqw?view#Implementation-Plan) this PR implements:
- Defining structure for ValidateFromExhaustive message
- Implements function `performBasicChecks` to confirm parameters are present and checks collator signature.
- Instantiates validation host.
- makes call to validate block of validation host to check the validation parameters and returns validation results message to caller.

Create tests to check for candidate validation exceptions based on  possible errors generated by validate block function and reviewing code from polkadot [test code](https://github.com/paritytech/polkadot-sdk/blob/b65313e81465dd730e48d4ce00deb76922618375/polkadot/node/core/candidate-validation/src/tests.rs#L396)

<!-- Brief list of functional changes -->

## Tests

<!-- Detail how to run relevant tests to the changes -->

```sh
 go test github.com/ChainSafe/gossamer/dot/parachain/candidate-validation -v
```

## Issues
closes: #3547 and #3995 
<!-- Write the issue number(s), for example: #123 -->

edwardmack committed:
create stubs to implement Subsystem interface

edwardmack committed:
create message types, move existing code to candidatevalidation package

edwardmack committed:
fix rebase conflicts

edwardmack committed:
added test wasm file

edwardmack committed:
address PR comments

edwardmack committed:
fix merge conflicts

edwardmack committed:
remove context from candidate validation subsystem, create message structs

edwardmack committed:
implement "happy path" functionality for validate from exhaustive

edwardmack committed:
implement perform basic checks and added tests

edwardmack committed:
address PR comments, remove redundant structs, cleanup

edwardmack committed:
address rebase conflicts, implement ValidationResultMessage

edwardmack committed:
add test for performBasicChecks function

edwardmack committed:
add test helper functions and additional tests

edwardmack committed:
cleanup comments

edwardmack committed:
fix message test to init validation host

edwardmack committed:
fix lint issues

edwardmack committed:
add chore use os.MkdirTemp + t.Cleanup fix (from PR#4037)

axaysagathiya commented on dot/parachain/candidate-validation/messages.go:
```diff
@@ -27,18 +29,19 @@ type ValidateFromExhaustive struct {
 
 // ValidationResult represents the result coming from the candidate validation subsystem.
-type ValidationResult struct {
+type ValidationResultMessage struct {
```
As you have renamed the struct, update the struct name in a comment.

edwardmack replied:
As per our discussion I've decided not to rename this.

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -74,7 +85,19 @@ func (cv *CandidateValidation) processMessages(wg *sync.WaitGroup) {
+				if err != nil {
+					logger.Errorf("failed to validate from exhaustive: %w", err)
+					msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResultMessage]{Err: err}
+					continue
```
We don't have any code in a for loop after the `select` block. Still, do we need the `continue` statement here?

edwardmack replied:
Removed, replaced with else condition.

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -74,7 +85,19 @@ func (cv *CandidateValidation) processMessages(wg *sync.WaitGroup) {
+					msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResultMessage]{Err: err}
+					continue
+				}
+				msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResultMessage]{
+					Data: ValidationResultMessage{
+						ValidationResult: *result,
+					},
+				}
```
- Validation results can be either `Valid` or `Invalid`. 

- If the result is `Invalid`,
    - set the `IsValid` field of `ValidationResultMessage` to false.
    - also store the reason for invalidity in the `Err` field of `ValidationResultMessage`.

- If the result is `Valid`,
    - set the `IsValid` field of `ValidationResultMessage` to true.
    - set the values of the `CandidateCommitments` and `PersistedValidationData` fields of `ValidationResultMessage`.
    
Reference from rust: https://github.com/paritytech/polkadot-sdk/blob/7ca0d65f19497ac1c3c7ad6315f1a0acb2ca32f8/polkadot/node/primitives/src/lib.rs#L346-L353

edwardmack replied:
I've updated this to assign the IsValid field and other fields in ValidationResult.

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +235,67 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
+	persistedValidationData parachaintypes.PersistedValidationData,
+	validationCode parachaintypes.ValidationCode,
+	candidateReceipt parachaintypes.CandidateReceipt, pov parachaintypes.PoV) (
+	*parachainruntime.ValidationResult, error) {
```
You are returning an incorrect type. It should be `ValidationResultMessage.`

edwardmack replied:
Updated function to return ValidationResult of parachain type.

axaysagathiya reviewed (changes requested):
the function returns the wrong struct.

axaysagathiya commented on dot/parachain/candidate-validation/messages.go:
```diff
@@ -27,18 +29,19 @@ type ValidateFromExhaustive struct {
 	IsValid                 bool
 	CandidateCommitments    parachaintypes.CandidateCommitments
 	PersistedValidationData parachaintypes.PersistedValidationData
+	ValidationResult        parachainruntime.ValidationResult
```
Why do we need this?

edwardmack replied:
We don't.  I've refactored the code to remove this.

edwardmack committed:
refactor ValidationResultMessage to use VDT

axaysagathiya commented:
We don't need to Encode or decode `ValidationResultMessage`, so we don't need to create VDT.

I have already commented above on how to use this struct.


> 
>     Validation results can be either Valid or Invalid.
> 
>     If the result is Invalid,
>         set the IsValid field of ValidationResultMessage to false.
>         also store the reason for invalidity in the Err field of ValidationResultMessage.
> 
>     If the result is Valid,
>         set the IsValid field of ValidationResultMessage to true.
>         set the values of the CandidateCommitments and PersistedValidationData fields of ValidationResultMessage.
> 
> Reference from rust: https://github.com/paritytech/polkadot-sdk/blob/7ca0d65f19497ac1c3c7ad6315f1a0acb2ca32f8/polkadot/node/primitives/src/lib.rs#L346-L353
> 

Let's schedule a meeting and work on this together if you are confused.

edwardmack commented:
> We don't need to Encode or decode `ValidationResultMessage`, so we don't need to create VDT.

I've removed the usage VDT for this type (and renamed it back to ValidationResult)

edwardmack committed:
remove the usage of VDT type for ValidationResult

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +258,83 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
+
+	validationCodeHash := validationCode.Hash()
+	// basic checks
+	err := performBasicChecks(&candidateReceipt.Descriptor, persistedValidationData.MaxPovSize, pov,
+		validationCodeHash)
+	if err != nil {
+		return &ValidationResult{ //nolint:nilerr  // note: we don't want to return an error here as the error is already set
+			IsValid: false,
+			Err:     err,
+		}, nil
```
Can't set this `err` as a reason for invalidity, 

`performBasicChecks` not only returns the reason for invalidity, but it also returns internal errors like hashing errors and encoding errors.

P1sar replied:
Yeah,  I do not see reason of returning whole structure when err could be enought to understand that structure is not valid, if you need exactly undrstand this, there could be 3rd bool param in return params

axaysagathiya replied:
@P1sar `ValidationResult` is the type we send to other subsystems. So, we need to return this struct and an IsValid field for more readability. 

- But yeah, we can remove the `IsValid` field. 
---
@edwardmack, we can return multiple errors from `performBasicChecks`. one for a reason for invalidity and the other for internal errors. (add a comment explaining return types).

timwu20 replied:
I don't see the need to keep an error in both `ValidationResult` and `parachaintypes.OverseerFuncRes`.  What is the difference to the receiver of this `OverseerFuncRes` type over a channel whether the error is within the `ValidationResult` or the `OverseerFuncRes` error.  I would assume you can omit the error `ValidationResult` type completely and just send back an `OverseerFuncRes[bool]` type to represent validity (IsValid) and an error.  You can leave it up to the function caller/consumer of the message to determine next steps based on the error returned using `errors.Is/As`

edwardmack replied:
@axaysagathiya I'll update `performBasciChecks` to return multiple errors.  What are your thought regarding removing err from `ValidationResult` (See comment below)?

axaysagathiya replied:
@edwardmack @timwu20 
- An error field in the `ValidationResult` is the reason for invalidity. 
(we can rename it to `reasonForInvalidity` for more readability) and 
- An error in `OverseerFuncRes` is an internal error


For more context,
> Validation results can be either Valid or Invalid.
> 
> If the result is Invalid,
>     set the IsValid field of ValidationResultMessage to false.
>     also store the reason for invalidity in the Err field of ValidationResultMessage.
> 
> If the result is Valid,
>     set the IsValid field of ValidationResultMessage to true.
>     set the values of the CandidateCommitments and PersistedValidationData fields of ValidationResultMessage.
> 

We have a dedicated structure for validation results, and the reason for Invalidity is a part of the result, so we should have this field in `ValidationResult`.

timwu20 replied:
I understand that's what you have now, but I don't see the reason to have different errors in different structures.  Why does the caller of this function through the overseer care if the error came from within `ValidationResult` or `OverseerFuncRes` error? The fact of the matter is, you will never return both a reason for invalidity and a system error.  It's either one or the other.  The caller can discern whether it's something like an error to be handled normally, or a systemic error which means that it shouldn't pass any more messages to the affected subsystem or even trigger it's own graceful shutdown.

axaysagathiya replied:
Again, `ValidationResult` stores the validation result, and the `reason for invalidity` is also a result.

- If a candidate is valid, we store some data in `ValidationResult.`
- So, in my opinion, we should also store the `reason for invalidity` in the `ValidationResult` only. (also, it is more readable)

Still, if you want, we can remove `reason for invalidity` from `ValidationResult.`

edwardmack replied:
I've updated the code for `preformBasicChecks`  and renamed the error in `ValidationResult`.  Let me know your thought everyone.

timwu20 committed:
fix candidate validation test

timwu20 commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -74,7 +85,42 @@ func (cv *CandidateValidation) processMessages(wg *sync.WaitGroup) {
+					msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResult]{
+						Data: *result,
+					}
+					//if err != nil {
+					//	logger.Errorf("failed to validate from exhaustive: %w", err)
+					//	msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResult]{
+					//		Data: ValidationResult{IsValid: false, Err: err},
+					//		Err:  nil,
+					//	}
+					//} else {
+					//	msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResult]{
+					//		Data: ValidationResult{
+					//			IsValid: true,
+					//			CandidateCommitments: parachaintypes.CandidateCommitments{
+					//				UpwardMessages:            result.UpwardMessages,
+					//				HorizontalMessages:        result.HorizontalMessages,
+					//				NewValidationCode:         result.NewValidationCode,
+					//				HeadData:                  result.HeadData,
+					//				ProcessedDownwardMessages: result.ProcessedDownwardMessages,
+					//				HrmpWatermark:             result.HrmpWatermark,
+					//			},
+					//			PersistedValidationData: msg.PersistedValidationData,
+					//			Err:                     nil,
+					//		},
+					//	}
```
is this supposed to be commented out?

edwardmack replied:
Yes, it's been removed (was leftover test code).

edwardmack committed:
remove un-used commented code

axaysagathiya commented:
> We don't need to Encode or decode `ValidationResultMessage`, so we don't need to create VDT.
> 
> I have already commented above on how to use this struct.
> 
> > ```
> > Validation results can be either Valid or Invalid.
> > 
> > If the result is Invalid,
> >     set the IsValid field of ValidationResultMessage to false.
> >     also store the reason for invalidity in the Err field of ValidationResultMessage.
> > 
> > If the result is Valid,
> >     set the IsValid field of ValidationResultMessage to true.
> >     set the values of the CandidateCommitments and PersistedValidationData fields of ValidationResultMessage.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Reference from rust: https://github.com/paritytech/polkadot-sdk/blob/7ca0d65f19497ac1c3c7ad6315f1a0acb2ca32f8/polkadot/node/primitives/src/lib.rs#L346-L353
> 
> Let's schedule a meeting and work on this together if you are confused.

Can you please add this comment to `ValidationResult` so that others can understand this structure?

edwardmack committed:
updated `performBasicChecks` to return multiple errors, rename error field in ValidationResult

edwardmack committed:
add comments

timwu20 commented on dot/parachain/candidate-validation/messages.go:
```diff
@@ -31,14 +32,24 @@ type ValidateFromExhaustive struct {
 	CandidateCommitments    parachaintypes.CandidateCommitments
 	PersistedValidationData parachaintypes.PersistedValidationData
-	Err                     error
+	ReasonForInvalidity     error
```
Given this is the actual type:
```
pub enum ValidationResult {
	/// Candidate is valid. The validation process yields these outputs and the persisted
	/// validation data used to form inputs.
	Valid(CandidateCommitments, PersistedValidationData),
	/// Candidate is invalid.
	Invalid(InvalidCandidate),
}
```
Why are we trying to store both the `Valid(CandidateCommitments, PersistedValidationData)` tuple and `Invalid(InvalidCandidate)` enum in the same struct? Couldn't we separate it out into separate types?  I was thinking something alone the lines of 
```
type ValidationResult interface {
	isValidationResult()
}
type ValidValidationResult struct {
	CandidateCommitments    parachaintypes.CandidateCommitments
	PersistedValidationData parachaintypes.PersistedValidationData
}

func (ValidValidationResult) isValidationResult() {}

type InvalidValidationResult struct {
	InvalidCandidate
}

func (InvalidValidationResult) isValidationResult() {}

type InvalidCandidate interface {
	isInvalidCandidate()
}
type ExecutionError string

func (ExecutionError) isInvalidCandidate()

type InvalidOutputs struct{}

func (InvalidOutputs) isInvalidCandidate()

type Timeout struct{}

func (Timeout) isInvalidCandidate()
... more enum options
```
That way we're not conflating errors with reason for an invalid candidate.  You can still implement the error interface for any of the `InvalidCandidate` rust variants/interface implementers if you choose to.  Or just treat it as an `InvalidCandidate`.

It'll be easy to do a type switch between `ValidValidationResult` and `InvalidValidationResult`.

kishansagathiya replied:
> That way we're not conflating errors with reason for an invalid candidate

We could solve that by having an enum for Candidate Invalidity details. So, ReasonForInvalidity could be that enum instead of just error.

https://github.com/paritytech/polkadot-sdk/blob/686aa233e67c619bcdbc8b758a9ddf92c3315cf1/polkadot/node/primitives/src/lib.rs#L324


> It'll be easy to do a type switch between ValidValidationResult and InvalidValidationResult.

Agreed. We could have a type for both of those things.

I don't like the approach of having `ValidationResult` and `InvalidCandidate` interface, though. I feel that approach could bloat the code and make it less readable. We have used such approaches at places. Hopefully, we will iterate through that and replace with some better approach.

edwardmack replied:
I've updated this to add types `ValidValidationResult` and `InvalidValidatiionResult` that will be used in `ValidationResult`.  Let me know your thoughts.

kishansagathiya commented on dot/parachain/types/types.go:
```diff
@@ -411,6 +415,15 @@ type HeadData struct {
 	Data []byte `scale:"1"`
 }
 
+func (hd HeadData) Hash() (common.Hash, error) {
```
Is this being used? I was able to remove this method and run `gossamer`, `lint` and `test` make targets

edwardmack replied:
I've removed this.

edwardmack replied:
Actually this is used in `dot/parachain/candidate-validation/candidate_validation.go:247` and `dot/parachain/candidate-validation/candidate_validation_test.go:531`

kishansagathiya replied:
> Actually this is used in dot/parachain/candidate-validation/candidate_validation.go:247 and dot/parachain/candidate-validation/candidate_validation_test.go:531

Code points that you mentioned uses validationCode.Hash(), this function is for HeadData.Hash()

edwardmack replied:
I see I checked the wrong function,  I have removed this now.

kishansagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +236,85 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
+
+// performBasicChecks Does basic checks of a candidate. Provide the encoded PoV-block.
+// Returns validation error or internal error if any.
+func performBasicChecks(candidate *parachaintypes.CandidateDescriptor, maxPoVSize uint32,
```
`validateFromChainState` also uses a similar set of basic check. Let's try to reuse this function in there.

edwardmack replied:
Yes I plan to do that when issue #3919 is addressed.

kishansagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +236,85 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
 }
+
+// validateFromExhaustive validates a candidate parachain block with provided parameters
+func validateFromExhaustive(validationHost parachainruntime.ValidationHost,
```
+1

edwardmack committed:
Add type ValidValidationResult and InvalidValidationResult to by used by ValidationResult struct

kishansagathiya commented on dot/parachain/candidate-validation/messages.go:
```diff
@@ -31,14 +32,27 @@ type ValidateFromExhaustive struct {
 }
 
-// PreCheck try to complie the given validation code and return the result
+type InvalidValidationResult struct {
+	ReasonForInvalidity error
+}
```
Not like this. You did not make candidate invalidity error an enum. Any error could be passed in ReasonForInvalidity.

Below is how to restrict this using an error enum. `CandidateInvalidity` is an int, but also an error since it implements the error interface.


```
type CandidateInvalidity int 

const (
	InvalidOutputs    CandidateInvalidity = iota
	Timeout
	ParamsTooLarge
	CodeTooLarge
	//...
)

func (ic CandidateInvalidity) Error() string {
	switch ic {
	case InvalidOutputs:
		return "validation outputs check doesn't pass"
	case Timeout:
		return "execution timeout"
	case ParamsTooLarge:
		return "validation input is over the limit"
	case CodeTooLarge:
		return "code size is over the limit"
	//...
	}

	return "unknown invalidity reason"
}
```

Extend this code for all the errors mentioned in https://github.com/paritytech/polkadot-sdk/blob/686aa233e67c619bcdbc8b758a9ddf92c3315cf1/polkadot/node/primitives/src/lib.rs#L324

edwardmack replied:
I've updated this to make candidate invalidity an enum.

edwardmack committed:
refactor ValidationResult InvaledResult as an enum to control possable errors

edwardmack committed:
Merge branch 'feat/parachain' into ed/feat/candidate_validation_exhaustive

kishansagathiya reviewed (approved):
epic! 🎉

axaysagathiya commented on dot/parachain/candidate-validation/messages.go:
```diff
@@ -31,14 +32,92 @@ type ValidateFromExhaustive struct {
 }
 
-// PreCheck try to complie the given validation code and return the result
+type CandidateInvalidity int
```
- Can we rename this type to `ReasonForInvalidity`?
- Let's change the datatype to `byte`.

edwardmack replied:
Renamed to `ReasonForInvalidity` and changed to byte.

axaysagathiya commented on dot/parachain/candidate-validation/messages.go:
```diff
@@ -31,14 +32,92 @@ type ValidateFromExhaustive struct {
 // ValidationResult represents the result coming from the candidate validation subsystem.
+// Validation results can be either a ValidValidationResult or InvalidValidationResult.
+//
+// If the result is invalid,
+// store the reason for invalidity in the Err field of InvalidValidationResult.
```
```suggestion
// If the result is invalid,
// store the reason for invalidity in the InvalidResult field of ValidationResult.
```

edwardmack replied:
Updated.

axaysagathiya commented on dot/parachain/candidate-validation/messages.go:
```diff
@@ -31,14 +32,92 @@ type ValidateFromExhaustive struct {
+// If the result is invalid,
+// store the reason for invalidity in the Err field of InvalidValidationResult.
+//
+// If the result is valid,
+// set the values of the CandidateCommitments and PersistedValidationData fields of ValidValidationResult.
```
```suggestion
// If the result is valid,
// set the values of the ValidResult field of ValidValidationResult.
```

edwardmack replied:
Updated.

axaysagathiya commented on dot/parachain/types/types.go:
```diff
@@ -640,6 +644,20 @@ type PoV struct {
+	return common.Blake2bHash(bytes)
+}
+
+func (p PoV) Encode() ([]byte, error) {
+	return scale.Marshal(p)
+}
```
Why do we need the Encode method?
Can't we directly marshal the POV wherever we need it?

edwardmack replied:
Yes, I've removed this.

edwardmack committed:
rename objects to address PR comments

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -74,7 +82,20 @@ func (cv *CandidateValidation) processMessages(wg *sync.WaitGroup) {
+					msg.ValidationCode, msg.CandidateReceipt, msg.PoV)
+				if err != nil {
+					logger.Errorf("failed to validate from exhaustive: %w", err)
+					msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResult]{
+						Data: *result,
+						Err:  err,
+					}
```
If there is validation error, Data should be nil
```suggestion
					msg.Ch <- parachaintypes.OverseerFuncRes[ValidationResult]{
						Err:  err,
					}
```

edwardmack replied:
Updated.

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +233,90 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
+	validationErr, internalErr := performBasicChecks(&candidateReceipt.Descriptor, persistedValidationData.MaxPovSize,
+		pov,
+		validationCodeHash)
+	if validationErr != nil {
+		validationResult := &ValidationResult{
+			InvalidResult: validationErr,
+		}
+		return validationResult, internalErr
+	}
```
Let's check for the internal error before the validation error.

What is the internal error?
 - example: encoding error, hashing error
 - If we get an internal error, that should be checked first, and internal errors should be returned.
 - if we get an internal error, it means we are not able to find the validation result.
 - So, we should return **either** a validation result **or** an internal error
 
```suggestion

	if internalErr != nil {
		return nil, fmt.Errorf("performing basic checks: %w", internalErr)
	}

	if validationErr != nil {
		validationResult := &ValidationResult{
			InvalidResult: validationErr,
		}
		return validationResult, nil
	}
```

edwardmack replied:
That makes more sense, I've updated this.

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +233,90 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
+	validationResult, err := validationHost.ValidateBlock(validationParams)
+	if err != nil {
+		ci := ExecutionError
+		return &ValidationResult{InvalidResult: &ci}, fmt.Errorf("executing validate_block: %w", err)
```
As I explained above, We should return either an internal error or a validation result.

edwardmack replied:
Updated

axaysagathiya commented:
It seems you have missed implementing [this](https://github.com/axaysagathiya/my-polkadot-sdk-old/blob/master/polkadot/node/core/candidate-validation/src/lib.rs#L667C4-L689C13).

edwardmack commented:
> It seems you have missed implementing [this](https://github.com/axaysagathiya/my-polkadot-sdk-old/blob/master/polkadot/node/core/candidate-validation/src/lib.rs#L667C4-L689C13).

Good point, I've implemented this now.

edwardmack committed:
add hash checks to validate candidate from exhaustive

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +232,110 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
+
+	validationResultHash, err := validationResult.HeadData.Hash()
+	if err != nil {
+		return nil, fmt.Errorf("hashing validation result: %w", err)
```
rename `validationResultHash` to `headDataHash`. Also, change the error string.

```suggestion
		return nil, fmt.Errorf("hashing head data: %w", err)
```

edwardmack replied:
updated.

axaysagathiya replied:
You have missed updating the variable from `validationResultHash` to `headDataHash`.

edwardmack replied:
Updated.

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -211,3 +232,110 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
+		return nil, fmt.Errorf("hashing validation result: %w", err)
+	}
+
+	if validationResultHash != candidateReceipt.Descriptor.ParaHead {
+		ci := ParaHeadHashMismatch
+		return &ValidationResult{InvalidResult: &ci}, nil
+	} else {
+		outputs := parachaintypes.CandidateCommitments{
+			UpwardMessages:            validationResult.UpwardMessages,
+			HorizontalMessages:        validationResult.HorizontalMessages,
+			NewValidationCode:         validationResult.NewValidationCode,
+			HeadData:                  validationResult.HeadData,
+			ProcessedDownwardMessages: validationResult.ProcessedDownwardMessages,
+			HrmpWatermark:             validationResult.HrmpWatermark,
+		}
+
+		// if validation produced a new set of commitments, we treat the candidate as invalid
+		if candidateReceipt.CommitmentsHash != outputs.Hash() {
+			ci := CommitmentsHashMismatch
+			return &ValidationResult{InvalidResult: &ci}, nil
+		} else {
+			return &ValidationResult{
+				ValidResult: &ValidValidationResult{
+					CandidateCommitments:    outputs,
+					PersistedValidationData: persistedValidationData,
+				},
+			}, nil
+		}
+	}
```
- Let's rename `output` to `candidateCommitments`.
- Also simplify the code by removing unnecessary `else` blocks. That would be more readable, IMO

```suggestion
	if validationResultHash != candidateReceipt.Descriptor.ParaHead {
		ci := ParaHeadHashMismatch
		return &ValidationResult{InvalidResult: &ci}, nil
	}

	candidateCommitments := parachaintypes.CandidateCommitments{
		UpwardMessages:            validationResult.UpwardMessages,
		HorizontalMessages:        validationResult.HorizontalMessages,
		NewValidationCode:         validationResult.NewValidationCode,
		HeadData:                  validationResult.HeadData,
		ProcessedDownwardMessages: validationResult.ProcessedDownwardMessages,
		HrmpWatermark:             validationResult.HrmpWatermark,
	}

	// if validation produced a new set of commitments, we treat the candidate as invalid
	if candidateReceipt.CommitmentsHash != candidateCommitments.Hash() {
		ci := CommitmentsHashMismatch
		return &ValidationResult{InvalidResult: &ci}, nil
	}

	return &ValidationResult{
		ValidResult: &ValidValidationResult{
			CandidateCommitments:    candidateCommitments,
			PersistedValidationData: persistedValidationData,
		},
	}, nil
```

edwardmack replied:
updated.

axaysagathiya reviewed (commented):
How do you plan to handle [these errors](https://github.com/axaysagathiya/my-polkadot-sdk-old/blob/7ca0d65f19497ac1c3c7ad6315f1a0acb2ca32f8/polkadot/node/core/candidate-validation/src/lib.rs#L633C1-L665C5)?

You have returned all the errors, but Rust doesn't consider them all errors. Two are considered errors, and others are reasons for invalidity.

timwu20 reviewed (approved):
Great work Ed!

edwardmack committed:
clearify code, rename varibale and simplify if/else condition

edwardmack committed:
rename variable

edwardmack commented:
> How do you plan to handle [these errors](https://github.com/axaysagathiya/my-polkadot-sdk-old/blob/7ca0d65f19497ac1c3c7ad6315f1a0acb2ca32f8/polkadot/node/core/candidate-validation/src/lib.rs#L633C1-L665C5)?
> 
> You have returned all the errors, but Rust doesn't consider them all errors. Two are considered errors, and others are reasons for invalidity.

The `ValidateBlock` function that creates these errors will be refactored when the PVF host is implemented, so we can parse the errors that are generated by that when it's implemented.

axaysagathiya commented:
> > How do you plan to handle [these errors](https://github.com/axaysagathiya/my-polkadot-sdk-old/blob/7ca0d65f19497ac1c3c7ad6315f1a0acb2ca32f8/polkadot/node/core/candidate-validation/src/lib.rs#L633C1-L665C5)?
> > You have returned all the errors, but Rust doesn't consider them all errors. Two are considered errors, and others are reasons for invalidity.
> 
> The `ValidateBlock` function that creates these errors will be refactored when the PVF host is implemented, so we can parse the errors generated by that function when it's implemented.

If we are not checking that the error is an actual error or reason for invalidity, it's not the correct implementation.
Can we add a to-do comment where errors should be filtered out?

Context:
Reason for invalidity: When a candidate is `Invalid,` we store the reason in the `validationResult` struct and don't consider it an error.

@kishansagathiya How should we handle errors here?

kishansagathiya commented:
> If we are not checking that the error is an actual error or reason for invalidity, it's not the correct implementation.
Can we add a to-do comment where errors should be filtered out?

@edwardmack 
Me and @axaysagathiya sat down to figure this out how to approach this. @edwardmack I agree with you that we could do this as a separate task. 

So, currently (I am including this PR), eventhough we have Reason for Invalidity (https://github.com/ChainSafe/gossamer/blob/2afed8b30a09fe2d44af649f25f059413d77ef84/dot/parachain/candidate-validation/messages.go#L56) we never use it.
I would think that ValidateBlock (https://github.com/ChainSafe/gossamer/blob/2afed8b30a09fe2d44af649f25f059413d77ef84/dot/parachain/runtime/instance.go#L69) should return a reason for invalidity in validation result. Right now it doesn't since ValidationResult struct (https://github.com/ChainSafe/gossamer/blob/2afed8b30a09fe2d44af649f25f059413d77ef84/dot/parachain/runtime/instance.go#L25) does not have any field in it for reason for invality or any sort of error. 

Let's make sure that ValidateBlock is implemented as expected. Let's have a decent number of tests for it. And differentiate between ReasonForInvalidity and errors occurred in executing ValidateBlock or any other internal error.

Have created an issue to track this task

edwardmack committed:
Merge branch 'feat/parachain' into ed/feat/candidate_validation_exhaustive
